deepmd/pt/utils/tabulate.py

```
@@ -0,0 +1,971 @@
+    def _convert_numpy_to_tensor(self):
+        """Convert self.data from np.ndarray to torch.Tensor."""
+        for ii in self.data:
+            self.data[ii] = torch.tensor(self.data[ii])  # pylint: disable=no-explicit-device&dtype
+
+    @property
+    @lru_cache
+    def _n_all_excluded(self) -> int:
+        """Then number of types excluding all types."""
+        return sum(int(self._all_excluded(ii)) for ii in range(0, self.ntypes))
+
+    @lru_cache
```

Ruff

Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks

(B019)